File: app/src/main/java/com/todaydiary/app/data/DiaryBodyDedupe.kt
package com.todaydiary.app.data

import com.todaydiary.app.ui.models.DiaryEntry
import java.time.LocalDate

/**
 * 불러오기 2회 / 저장 차이로 생기는 **미세한 공백·개행** 차이를 없앤 뒤 같은 일기인지 본다.
 */
object DiaryBodyDedupe {
    fun normForDedupe(body: String): String =
        body
            .replace("\r\n", "\n")
            .replace('\r', '\n')
            .replace(Regex("""\s+"""), " ")
            .trim()

    private fun key(date: LocalDate, body: String) = date to normForDedupe(body)

    /** 같은 날 + [normForDedupe]로 같으면 1건만(사진·id 우선) */
    fun pickOnePerDuplicateKey(entries: List<DiaryEntry>): List<DiaryEntry> {
        if (entries.size < 2) return entries
        return entries
            .groupBy { key(it.date, it.body) }
            .map { (_, group) ->
                if (group.size == 1) {
                    group.first()
                } else {
                    group.maxWith(
                        compareByDescending<DiaryEntry> { it.photos.size }
                            .thenBy { it.id }
                    )
                }
            }
    }
}

File: web/src/features/diary/diaryBodyDedupe.ts
import type { DiaryEntry } from './types';

/** Android `DiaryBodyDedupe.normForDedupe` */
export function normForDedupe(body: string): string {
  return body
    .replace(/\r\n/g, '\n')
    .replace(/\r/g, '\n')
    .replace(/\s+/g, ' ')
    .trim();
}

/** Android `DiaryBodyDedupe.pickOnePerDuplicateKey` */
export function pickOnePerDuplicateKey(entries: DiaryEntry[]): DiaryEntry[] {
  if (entries.length < 2) return entries;
  const groups = new Map<string, DiaryEntry[]>();
  for (const e of entries) {
    const k = `${e.date}|${normForDedupe(e.body)}`;
    const g = groups.get(k) ?? [];
    g.push(e);
    groups.set(k, g);
  }
  const out: DiaryEntry[] = [];
  for (const group of groups.values()) {
    if (group.length === 1) {
      out.push(group[0]);
    } else {
      out.push(
        group.reduce((best, cur) => {
          if (cur.photos.length > best.photos.length) return cur;
          if (cur.photos.length < best.photos.length) return best;
          return cur.id > best.id ? cur : best;
        }),
      );
    }
  }
  return out;
}
